### ISSUE TITLE
offlineManager.getPacks() crashes app

### ISSUE THREAD

**Issue by aaronblondeau** [nitaliano/react-native-mapbox-gl#907]
This occurs on both iOS and Android with version 6.0.2.

Steps to reproduce:
1) Create a brand new react native project with "react-native init"
2) Install the mapbox gl package (I use the manual install steps for both iOS and Android)
3) Use the following App.js

```
import React, { Component } from 'react';
import {
  Platform,
  StyleSheet,
  Text,
  View,
  Button
} from 'react-native';

import Mapbox from '@mapbox/react-native-mapbox-gl';
Mapbox.setAccessToken('YOUR TOKEN HERE');

async function breakStuff () {
  var packs = await Mapbox.offlineManager.getPacks()
}
breakStuff()

export default class App extends Component<{}> {
  render() {
    return (
      <View style={styles.container}>
        <Mapbox.MapView
            styleURL={'mapbox://styles/salidasoftware/cjafzjpen7l322spl16u1y6u5'}
            zoomLevel={15}
            centerCoordinate={[11.256, 43.770]}
            style={styles.container}>
        </Mapbox.MapView>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});
```

Android will always crash.  iOS may crash immediately, or you may have to stop and re-run the app and it will crash every time after that.

iOS Crash is : com.mapbox.mbgl.DefaultFileSource (25): EXC_BAD_ACCESS (code=1, address=0x0)
Android Crash is : Fatal signal 11 (SIGSEGV), code 1, fault addr 0x0 in tid 27552 (GLThread 763)

I am trying to manage the offline packs in a separate module, but distilled the crash down to the example above.

**Comment by nitaliano** [comment]
Try calling the breakStuff in componentWillMount instead of outside of the React component. setAccessToken is asynchronous by nature because the call goes over the bridge, so there may be a timing issue

**Comment by aaronblondeau** [comment]
I moved the getPacks call inside MapView's onDidFinishLoadingMap and everything is working on iOS.  Android no longer crashes, but shows a black map after that first load. Still searching for a workaround there...

**Comment by aaronblondeau** [comment]
I've been digging around in the Java code to try and find a way to fix this.  Once OfflineManager's listOfflineRegions method is called, no more map data gets loaded until the app is restarted.  I've added logging to the module's Java code to ensure that Mapbox.getAccessToken() returns the right token before the call is made, and it is.  I also tried reproducing this with a plain Android project and the Android SDK and cannot get it to happen there.  I wonder if the react context is breaking something inside of FileSource?

**Comment by nitaliano** [comment]
I wonder if calling `mReactContext.getApplicationContext() ` or ` mReactContext.getBaseContext()` [here](https://github.com/mapbox/react-native-mapbox-gl/blob/master/android/rctmgl/src/main/java/com/mapbox/rctmgl/modules/RCTMGLOfflineModule.java#L107) would change anything

**Comment by nitaliano** [comment]
@aaronblondeau did you happen to use the previous version of this repo in your application before we supported it? I'm asking because I'm wondering if it could be a weird db migration issue

**Comment by aaronblondeau** [comment]
@nitaliano - Using getApplicationContext() or getBaseContext() didn't help.  I was also making sure to do clean installs and even using fresh devices so not a migration issue.

I did finally find a way around this though.  Probably not a fix, but something that might help find the true cause.

Suspecting that FileSource was the cause I reduced getPacks down to just the FileSource code that was being called:

```
    @ReactMethod
    public void getPacks(final Promise promise) {
        FileSource fileSource = FileSource.getInstance(mReactContext);
        fileSource.activate();
        fileSource.deactivate();

        WritableArray payload = Arguments.createArray();
        promise.resolve(payload);
    }
```
Sure enough, this prevents tiles from loading in the map.  I then found that removing the call to fileSource.deactivate() would eliminate the problem, so I modified getPacks to this:

```
    @ReactMethod
    public void getPacks(final Promise promise) {
        
        // I added these two lines
        FileSource fileSource = FileSource.getInstance(mReactContext);
        fileSource.activate();

        final OfflineManager offlineManager = OfflineManager.getInstance(mReactContext);

        offlineManager.listOfflineRegions(new OfflineManager.ListOfflineRegionsCallback() {
            @Override
            public void onList(OfflineRegion[] offlineRegions) {
                WritableArray payload = Arguments.createArray();

                for (OfflineRegion region : offlineRegions) {
                    payload.pushMap(fromOfflineRegion(region));
                }

                promise.resolve(payload);
            }

            @Override
            public void onError(String error) {
                promise.reject("getRegions", error);
            }
        });

    }
```

And now the map will continue to load tiles after a call to getPacks.  The listOfflineRegions call will still call fileSource.deactivate() so nothing should get left open.  Very puzzling why things would break in the original workflow though.

**Comment by alexgvozden** [comment]
is there any official fix for this as I am getting same crashes EXC_BAD_ACCESS on iOS and crashes on Android (not all the time)

<img width="859" alt="screen shot 2018-01-12 at 4 16 54 pm" src="https://user-images.githubusercontent.com/127109/34881677-7de08ad6-f7b4-11e7-9a39-7d71ab7dd57c.png">

**Comment by nitaliano** [comment]
I'll take a look at it today. I'll also try updating to the latest Android and iOS SDKs to see if that helps

Update: I believe I fixed the crashes on iOS so I'll be sending out a PR for that today

@aaronblondeau I can verify that what you suggested is working, one thing that I did notice is that in  the Android SDK `OfflineManger.class` the constructor is grabbing the file source only once and reusing it throughout the class

```
private OfflineManager(Context context) {
        this.context = context.getApplicationContext();
        this.fileSource = FileSource.getInstance(this.context);
        this.initialize(this.fileSource);
        this.deleteAmbientDatabase(this.context);
    }

public void listOfflineRegions(@NonNull final OfflineManager.ListOfflineRegionsCallback callback) {
        this.fileSource.activate();
        this.listOfflineRegions(this.fileSource, new OfflineManager.ListOfflineRegionsCallback() {
            public void onList(final OfflineRegion[] offlineRegions) {
                OfflineManager.this.getHandler().post(new Runnable() {
                    public void run() {
                        OfflineManager.this.fileSource.deactivate();
                        callback.onList(offlineRegions);
                    }
                });
            }

            public void onError(final String error) {
                OfflineManager.this.getHandler().post(new Runnable() {
                    public void run() {
                        OfflineManager.this.fileSource.deactivate();
                        callback.onError(error);
                    }
                });
            }
        });
    }
```

but for some reason when we grab the file source before our calls into the Android SDK and activating ourselves the problem goes away, the React context is totally breaking this.

What is also interesting is the other methods inside the Android SDK activate the FileSource like this
```
FileSource.getInstance(this.context).activate();
```

listOfflineRegions is the only one that uses `this.fileSource.activate()`

**Comment by nitaliano** [comment]
PR is out if you guys wanted to test it out https://github.com/mapbox/react-native-mapbox-gl/pull/949

**Comment by aaronblondeau** [comment]
@nitaliano the package is no longer causing crashes for me on Android with offline packs - thanks!

**Comment by nitaliano** [comment]
This is fixed now and merged into master closing out
